docs: restructure working PR (#4866)
dfinity/motoko: doc/md/fundamentals/2-types/3-functions.md

---
sidebar_position: 3
---

# Function types

> Motoko provides different types of functions based on where in the program they execute and how they interact with the system. Understanding these distinctions is essential when designing efficient and scalable canister logic.

Reviewer: Like other languages, Motoko lets you define functions to abstract code into re-usable building blocks.
Functions are parameterized bits of code that, when supplied with arguments produce results. In the process
of producing a result, the execution of the function can have other side-effects, like modifying state, print to the log, or sending messages to other canisters.

Functions can be synchronous or asynchronous. A synchronous function blocks the caller until the function returns control with its result. allowing the caller to proceed with its result.

An asynchronous functions immediately returns to its caller - its execution is decoupled from the execution of the caller. Instead, the caller may receive a future, a placeholder for the result of the call. 
This future is completed with a result by the callee, when it returns. 
The caller can retrieve the result by awaiting the future at a later point in time. It can also ignore the result entirely by never awaiting the future.

> Motoko provides different types of functions based on where in the program they execute and how they interact with the system. Understanding these distinctions is essential when designing efficient and scalable canister logic.

Reviewer: ```suggestion
Like other languages, Motoko lets you define functions to abstract code into re-usable building blocks.
Functions are parameterized bits of code that, when supplied with arguments produce results. In the process
of producing a result, the execution of the function can have other side-effects, like modifying state, print to the log, or sending messages to other canisters.

Functions can be synchronous or asynchronous. A synchronous function blocks the caller until the function returns control with its result. allowing the caller to proceed with its result.

An asynchronous functions immediately returns to its caller - its execution is decoupled from the execution of the caller. Instead, the caller may receive a future, a placeholder for the result of the call. 
This future is completed with a result by the callee, when it returns. 
The caller can retrieve the result by awaiting the future at a later point in time. It can also ignore the result entirely by never awaiting the future. 
```

| `query`     | Optimized for reading data but cannot modify [state](https://internetcomputer.org/docs/motoko/fundamentals/state). |

## Function comparison

| Function type                | Mutates [state](https://internetcomputer.org/docs/motoko/fundamentals/state) | Calls updates | Calls queries | Asynchronous | External calls |
|------------------------------|---------------|------------------|------------------|---------------|---------------|
| Local                    | No            | Yes              | Yes              | No            | No            |
| Generic                  | No            | Yes              | Yes              | No            | No            |
| Shared update            | Yes           | Yes              | Yes              | Yes           | Yes           |
| Shared query             | No            | No               | Yes              | Yes           | Yes           |
| Shared composite query   | No            | No               | Yes              | Yes           | Yes           |

## Local functions

> Local functions run within the canister's [actor](https://internetcomputer.org/docs/motoko/fundamentals/actors-async). They do not make calls to other [canisters](https://internetcomputer.org/docs/building-apps/essentials/canisters). Local functions are processed fast and synchronously.

Reviewer: ```suggestion
Local functions run within the canister's [actor](https://internetcomputer.org/docs/motoko/fundamentals/actors-async). They do not make calls to other [canisters](https://internetcomputer.org/docs/building-apps/essentials/canisters). Local functions are cheap to call and execute synchronously.
```

```motoko no-repl
func add(a : Nat, b : Nat) : Nat {
    a + b
};

let sum : Nat = add(5, 3); // Synchronous execution
```

> 

Reviewer: ```suggestion

The type of `add` is `(Nat, Nat) -> Nat` indicating that it expects a pair of naturals  as argument and returns a natural as a result.
```

> **Example use case:** Simple computations that do not require network calls.

Reviewer: ```suggestion
**Example use case:** Local computations that do not require communication with other actors or canisters.
```

## Generic functions

Generic functions allow the use of type parameters, making them more flexible for using different data types.

```motoko no-repl
func identity<T>(x : T) : T = x;

let num : Nat = identity<Nat>(42);
let text : Text = identity<Text>("Hello");

> ```

Reviewer: ````suggestion
```

The type of `identity` is `<T>T -> T`, indicating that when supplied with any type `T`, and argument of type `T`, it  returns a result of the same type `T`.

Type arguments can be omitted from calls when the compiler can infer them from the arguments and context, allowing the simpler code:

```motoko no-repl
func identity<T>(x : T) : T = x;
let num : Nat = identity(42);
let text : Text = identity("Hello");
```
````

> 

Reviewer: ```suggestion

## Local, asynchronous functions

Local function that have an `async`  or `async*` return type are asynchronous and can interact with other canisters by calling shared functions.

They are useful for defining  asynchronous logic  used in the implementation of public  `shared` functions.

``` motoko no-repl

import Time "mo:base/Time";
import Logger "canister:Logger";

actor {

    private func log(msg : Text) : async () {
        Logger.log(Time.now() + msg); // sends a message
    };

   public shared func doStuff() : async () {
       await log("doingStuff"); 
   } 
}
```

A more efficient variation is to use `async*` and `await*` that avoids the overhead of messaging just to call the local function:

``` motoko no-repl

import Time "mo:base/Time"
import Logger "canister:Logger";

actor {

    var logging = false;

    private func log(msg : Text) : async* () {
        if (logging) { 
           Logger.log(Time.now() + msg); // sends a message
        }
    };

   public shared func doStuff() : async () {
       await* log("doingStuff");
   } 
}
```

## Shared functions

> 

Reviewer: ```suggestion

The public functions of an actor determine its external interface. All public functions in an actor must be shared and can be either `shared`, `shared query` or `shared composite query` functions. Private functions cannot be `shared`. 
Since public functions must be `shared`, the `shared` keyword is optional and can be committed.

`shared` functions permanently update the state of an actor, while `query` and `composite query` functions are only executed for their result. While queries can temporarily alter the state of the actor, the changes are not permanent and never visible to other callers. It's as if each query operated on a copy of the actor that is discarded on return from the query. When called from a front-end, `query` functions have much lower latency than calling the equivalent `shared` method. This is because `shared` functions require the protocol to reach consensus on the state changes and result of a shared function, while `query` functions do not.
```

Shared functions can be called from outside the canister (e.g., by users or other [canisters](https://internetcomputer.org/docs/building-apps/essentials/canisters)). These are processed [asynchronously](https://internetcomputer.org/docs/motoko/fundamentals/actors-async#async--await) because they interact with the network.

> ```motoko no-repl
>   public shared func getBalance() : async Nat {
>     return 100;
>   };
> ```

Reviewer: ````suggestion
```motoko no-repl

persistent actor Account {
  var balance = 0;
  
  public shared func deposit(amount : Nat) : async Nat {
     balance += amount;
     balance
  };

}
```

Omitting the `shared` keyword, we can also write:

````suggestion
```motoko no-repl

persistent actor Account {
  var balance = 0;
  
  public shared func deposit(amount : Nat) : async Nat {
     balance += amount;
     balance
  };

}
```

The `deposit` function has type : `shared Nat -> async Nat`.

````

Its better to use an example that updates the state.

> **Example use case:** Reusable logic for different types.

Reviewer: ```suggestion
**Example use case:** Operations that must update the state of an actor, such as transactions.
```

## Query functions

> [Query](https://internetcomputer.org/docs/building-apps/interact-with-canisters/query-calls) functions are designed for retrieving data. They cannot modify [state](https://internetcomputer.org/docs/motoko/fundamentals/state) and execute faster than [update](https://internetcomputer.org/docs/building-apps/interact-with-canisters/update-calls) functions because they do not go through consensus. Query functions are identified with the `query` keyword. Any function without the `query` keyword is an [update](https://internetcomputer.org/docs/building-apps/interact-with-canisters/update-calls) function.

Reviewer: ```suggestion
[Query](https://internetcomputer.org/docs/building-apps/interact-with-canisters/query-calls) functions are designed for retrieving data. They cannot permanently update [state](https://internetcomputer.org/docs/motoko/fundamentals/state) and execute faster than [update](https://internetcomputer.org/docs/building-apps/interact-with-canisters/update-calls) functions because they do not go through consensus. Query functions are identified with the `query` keyword. Any function without the `query` keyword is an [update](https://internetcomputer.org/docs/building-apps/interact-with-canisters/update-calls) function.
```

> ```motoko no-repl
>   public query func greet(name : Text) : async Text {
>     return greeting # name # "!";
>   };
> ```
> 
> ### Shared query functions
> 
> ```motoko
> actor Bank {
>     stable var balance : Nat = 100;
> 
>     public query func getBalance() : async Nat {
>         return balance;
>     };
> };
> await Bank.getBalance();
> ```
> 
> **Example use case:** Fetching data quickly without modifying the canister [state](https://internetcomputer.org/docs/motoko/fundamentals/state).

Reviewer: ````suggestion
```motoko no-repl
persistent actor Account {
  var balance  = 0;
  public shared query func getBalance() : async Nat {
     balance
  };
}
```

Again, you can omit the `shared` keyword:

```motoko no-repl
persistent actor Account {  
  var balance = 0;

  public query func getBalance() : async Nat {
     balance
  };
}
```

The `getBalance` function has function type `query () -> async Nat`

**Example use case:** Fetching data quickly without modifying the canister [state](https://internetcomputer.org/docs/motoko/fundamentals/state).
````

```motoko no-repl
import Nat "mo:base/Nat";

actor Counter {
  stable var count : Nat = 0;
  public query func getCount() : async Nat {
    return count
  };
  public composite query func isEven() : async Bool {
    return (await getCount()) % 2 == 0
  };
  public composite query func summary() : async Text {
    let current = await getCount();
    let even = await isEven();
    if (even) {
      return "Count is " # Nat.toText(current) # " and it is even."
    } else {
      return "Count is " # Nat.toText(current) # " and it is odd."
    }
  };
}
```

### Shared composite query functions

```motoko no-repl
actor Bank {
    stable var balance : Nat = 100;

    public query func getBalance() : async Nat {
        return balance;
    };

    public composite query func doubleBalance() : async Nat {
        let b : Nat = await getBalance();
        return b * 2;
    };
};
await Bank.doubleBalance();
```

**Example use case:** Efficiently combining multiple queries while avoiding [inter-canister](https://internetcomputer.org/docs/motoko/fundamentals/messaging) update calls.

> ## Update functions
> 
> [Update](https://internetcomputer.org/docs/building-apps/interact-with-canisters/update-calls) functions modify a canister’s [state](https://internetcomputer.org/docs/motoko/fundamentals/state) and must go through consensus before the result is returned. Any function without the identifier `query` is an update function by default.
> 
> ```motoko no-repl
>   public func setGreeting(prefix : Text) : async () {
>     greeting := prefix;
>   };
> ```
> 
> **Example use case:** Exposing public endpoints for [inter-canister](https://internetcomputer.org/docs/motoko/fundamentals/messaging) or [frontend](/docs/building-apps/frontends/using-an-asset-canister) interactions.
> 
> ### Shared update functions
> 
> ```motoko
> actor Bank {
>     stable var balance : Nat = 0;
> 
>     public shared func deposit(amount : Nat) : async () {
>         balance += amount;
>     };
> };
> ```
> 
> **Example use case**: Transactions, user [state](https://internetcomputer.org/docs/motoko/fundamentals/state) updates, or anything that modifies persistent data.

Reviewer: ```suggestion
```

I think this is now redundant since I  moved the deposit example further up.

## Passing arguments to functions

An argument is a value passed to a function when called. Arguments enable functions to process incoming data during execution.

### Single argument

A function can take a single argument of a specific type.

```motoko no-repl
  public func increment(amount : Nat) : async Nat {
    count += amount;
    return count;
  }
```

> ### Multiple arguments

Reviewer: ```suggestion
### Multiple arguments and returns
```

Functions can accept multiple arguments by separating them with commas.

```motoko no-repl
  public func add(x : Nat, y : Nat) : async Nat {
    return x + y;
  }
```

### Using a record as an argument

Multiple values can be passed as a single argument by encapsulating them within a [record](https://internetcomputer.org/docs/motoko/fundamentals/types/records) type.

```motoko no-repl

> type UserData = { name : Text; age : Nat };

Reviewer: ```suggestion
  type UserData = { name : Text; age : Nat };
```

>   public func createUser(userData : UserData) : async Text {

Reviewer: ```suggestion
   func createUser(userData : UserData) : async Text {
```

> ```motoko no-repl
> type UserData = { name : Text; age : Nat };
>   public func createUser(userData : UserData) : async Text {
>     return userData.name;
>   }
> ```

Reviewer: ````suggestion
```motoko no-repl
  public func userName(user: { name : Text; age : Nat }) : Text {
    user.name;
  }
```
Or more simply using a record pattern:
```motoko no-repl
  public func userName({ name : Text; age : Nat }) : Text {
    name;
  }
```
